File: challenges/week6_stay_solid_friends/solutions/refactored.php
<?php

/**

## broken solid principle's:  OCP, SRP


class InvoicePrinter {
    public function printInvoice($invoice, $format) {
        if ($format == 'pdf') {
            echo "Printing PDF invoice...";
        } else if ($format == 'html') {
            echo "Printing HTML invoice...";
        } else {
            echo "Unknown format";
        }
    }
}
 */


///////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////
/*
 * refactored code addresses the following:
 *
 * SRP: InvoicePrinter class no long responsible for print pdfs and html
 * OCP: You can now add more formats)(json,text,etc) without 
 *      having to adjust the base
 */

abstract class InvoicePrinter {
    abstract public function printInvoice($invoice): void;
}

class PdfInvoice extends InvoicePrinter{
    public function printInvoice($invoice): void {
            echo "\nPrinting PDF invoice...\n";
    }
}


class HtmlInvoice extends InvoicePrinter{
    public function printInvoice($invoice): void {
            echo "\nPrinting HTML invoice...\n";
    }
}


// Example usage:
$invoice = ['id' => 123]; // some invoice data

$pdf = new PdfInvoice();
$pdf->printInvoice($invoice); 
$html = new HtmlInvoice();
$html->printInvoice($invoice); 

File: challenges/week6_stay_solid_friends/solutions/refactored.py
## broken solid principle hint (in hex): SRP, OCP

## Original code:
##
##class Report:
##    def __init__(self, data):
##        self.data = data
##
##    def generate(self):
##        return f"Report: {self.data}"
##
##    def save_to_file(self, filename):
##        with open(filename, "w") as f:
##            f.write(self.generate())


###########################################################
###########################################################
###########################################################
##
## refactored code addresses the following:
##
## SRP: Report now long responsible for generating report
##      and write the file somewhere
## OCP: We now can add a DatabaseWriter, CloudWriter, etc
##      and we wouldnt have to modify report 
##
###########################################################
###########################################################

class Report:
    def __init__(self, data):
        self.data = data

    def generate(self):
        return f"Report: {self.data}"


class FileWriter:
    def save_to_file(self, filename: str, content: str):
        with open(filename, "w") as f:
            f.write(content)
